syntax = "proto3";
package fc_pb;

message Controller {
    ControllerConfig config = 1;
    Devices devices = 2;
}

message ControllerConfig {
    uint32 update_interval = 1;
    bool dynamic = 2;
    uint32 smoothing_intervals = 3;
    uint32 top_stickiness_intervals = 4;
    uint32 temp_averaging_intervals = 5;
}

message Devices {
    repeated Fan fan = 1;
    repeated Sensor sensor = 2;
}

enum DevType {
    SYS = 0;
    DELL = 1;
    NVIDIA = 2;
}

message Fan {
    DevType type = 1;
    string label = 2;
    string sensor = 3;
    string temp_to_rpm = 4;
    string rpm_to_pwm = 5;
    uint32 start_pwm = 6;
    uint64 interval = 7;
    bool ignore = 8;

    // SYS & DELL
    int32 driver_flag = 10;
    string pwm_path = 11;
    string rpm_path = 12;
    
    // SYS
    string enable_path = 13;    

    // NV
    uint32 id = 20;
}

message FanLabel {
    string label = 1;
}

message Sensor {
    DevType type = 1;
    string label = 2;

    // SYS
    string input_path = 10;
    string enable_path = 11;
    string fault_path = 12;
    string min_path = 13;
    string max_path = 14;
    string crit_path = 15;

    // NV
    uint32 id = 20;
}

message Empty {}

message TestRequest {
    string device_label = 1;
    bool forced = 2;
}

message TestResponse {
    int32 status = 1;
}

message FanStatus {
    enum Status {
        ENABLED = 0;
        DISABLED = 1;
        TESTING = 2;
    }
    string label = 1;
    Status status = 2;
    uint32 rpm = 3;
    uint32 pwm = 4;
}

service DService {
    rpc StopService(Empty) returns (Empty) {}
    rpc GetDevices(Empty) returns (Devices) {}
    rpc SetDevices(Devices) returns (Empty) {}
    rpc SubscribeDevices(Empty) returns (stream Devices) {}
    rpc GetEnumeratedDevices(Empty) returns (Devices) {}
    rpc GetControllerConfig(Empty) returns (ControllerConfig) {}
    rpc SetControllerConfig(ControllerConfig) returns (Empty) {}
    
    rpc GetFanStatus(FanLabel) returns (FanStatus) {}
    rpc SubscribeFanStatus(Empty) returns (stream FanStatus) {}
    rpc Enable(FanLabel) returns (Empty) {}
    rpc EnableAll(Empty) returns (Empty) {}
    rpc Disable(FanLabel) returns (Empty) {}
    rpc DisableAll(Empty) returns (Empty) {}
    rpc Test(TestRequest) returns (stream TestResponse) {}
    rpc Reload(Empty) returns (Empty) {}
    rpc Recover(Empty) returns (Empty) {}
    rpc NvInit(Empty) returns (Empty) {}
}
